### ISSUE TITLE
enable `Variable width tile matrices` 

### ISSUE THREAD

**Issue by vincentsarago** [developmentseed/morecantile#115]
> Until now, it has been assumed that matrixWidth is constant for all tile rows. This is common usage for projections that do not distort the Earth too much. But when using Equirectangular Plate Carrée projection (see [Annex D](https://docs.ogc.org/is/17-083r4/17-083r4.html#annex-common-tilematrixset-definitions-informative) subsection 2) the distortion increases for tiles closer to the poles. In the extreme, the upper row of the upper tile (the one representing the North Pole) contains a list of repeated values that represents almost the same position in the space. The same can be said for the lower row of the lower tile (the one representing the South Pole). When the tiles are represented in a flat projection, this is an effect that cannot be avoided, but when the data are presented in a virtual globe, the distortion results in redundant information in the poles that need to be eliminated by the client during the rendering. Compensating for distortion is better done at the server side instead.

> The solution consists of reducing the number of tiles (matrixWidth) in the high latitude rows and generating those tiles with a compressed scale in the longitudinal dimension (see [Figure 8](https://docs.ogc.org/is/17-083r4/17-083r4.html#fig8)). To allow this solution, the tile model must be extended to specify coalescence coefficients (𝑐) that reduce the number of tiles in the width direction by aggregating 𝑐 horizontal tiles but keeping the tileWidth (and tileHeight). The coalescence coefficient is not applied next to the Equator but is used in medium and high latitudes (the higher the latitude the larger the coefficient).

> Even if tiles can coalesce, this does not change the indexing or the tile matrix set that will be the same as if no coalescence has been applied. For example, if the 𝑐 coefficient is 4, the tileCol of the first tile will be 0, the tileCol of the second tile will be 4, the tileCol of the third tile will be 8 and so on. In other words, and for the same example, tileCol 0, 1, 2 and 3 points to the same tile.

ref: https://docs.ogc.org/is/17-083r4/17-083r4.html#toc15

I've added the needed attributes in the model (commented) https://github.com/developmentseed/morecantile/blob/f346786d8e1164af255eb61dfc8c251859ca8cd1/morecantile/models.py#L116-L125 and https://github.com/developmentseed/morecantile/blob/f346786d8e1164af255eb61dfc8c251859ca8cd1/morecantile/models.py#L185 but I'm not quite sure I understand how it works 😓

**Comment by AndrewAnnex** [comment]
@vincentsarago this would actually be neat to see implemented. Ny looking at that section I basically see that depending on the row you just lower the matrixWidth by dividing it by the coalescence coefficient, eg if in top row and c =4 you would have width/4 martixWidth. I'm looking at figure 8 to understand this, but no idea if it really is as easy as dividing the matrixWidth within morecantile. It also looks like Figure E.1 has a figure about how to scale the coefficients to be closer to equal area, and figure E.3 has a good visual for how the tiles look in equirectangular projection and example confs

**Comment by jerstlouis** [comment]
@AndrewAnnex 

It's actually very simple.

For rows where a coalescing factor is set, that many tiles will be coalesced into a single tile, with the same number of pixels as usual, but covering the horizontal extent of all of these tiles.

As an example, for a row with c=2, two 256x256 tiles covering 45 degrees x 45 degrees become a single 256x256 tile covering 45 degrees of latitude by 90 degrees of longitude.

I will be participating in the code sprint and happy to help clarify and answering any questions.

**Comment by vincentsarago** [comment]
FYI I started a PR in #131 
I'll need to add some tests and also so if all the morecantile methods are working with VariableWidth TMS

**Comment by vincentsarago** [comment]
I've added some test and put the PR as ready to review.

if someone want to tests that will be awesome, else I'll just assume that I understood the spec correctly 😅 

```python
import morecantile
tms = morecantile.tms.get("GNOSISGlobalGrid")

tms.xy_bounds(0, 0, 0)
>> BoundingBox(left=-180.0, bottom=0.0, right=-90.0, top=90.0)

tms.xy_bounds(0, 0, 1)
>> BoundingBox(left=-180.0, bottom=45.0, right=-90.0, top=90.0)

tms.xy_bounds(1, 0, 1)
>> BoundingBox(left=-180.0, bottom=45.0, right=-90.0, top=90.0)
```

```
morecantile tms --identifier GNOSISGlobalGrid | morecantile tms-to-geojson --level 2 --collect | gist
```

https://gist.github.com/vincentsarago/36d1dafc34be6b2ba9fd926bae8f07fa

<img width="1003" alt="Screenshot 2023-07-21 at 6 27 03 PM" src="https://github.com/developmentseed/morecantile/assets/10407788/1c41892b-60cf-4204-98b4-c5d4485e02b9">

**Comment by jerstlouis** [comment]
> I think the spec says it exist but it has the same bounds as the tile 0,0,1. I think it might be just different wording.

Right. The important thing is that when generating a tileset (or when a client request tiless), you do not generate the alias tile as well as the "real" tile.
Otherwise that will completely defeat the purpose of variable width, since you are re-duplicating the same data :D

**Comment by vincentsarago** [comment]
> Right. The important thing is that when generating a tileset (or when a client request tiless), you do not generate the alias tile as well as the "real" tile.
> Otherwise that will completely defeat the purpose of variable width, since you are re-duplicating the same data :D

ah, I was more under the impression that it's up to the client to not call the tile 😬

Not sure about the tileset, if I remember well the tileset specification says you need provide min/max row/col so it's the same issue it's up to the client to know the tilematrixset and not `ask` for `alias` tiles.

**Comment by jerstlouis** [comment]
@vincentsarago

I was mainly mentioning this for the case of a an offline tileset batch producer (not sure whether morecantile is used for this or not), where you definitely don't want to produce those extra tiles as it will make the resulting tileset larger than it needs to be, which is a key thing that the VW TMS try to avoid.

For a server, ideally it would return a 3xx redirect to the correct tile. For an on-demand server / block storage, you ideally want to avoid extra cache usage or processing overhead from generating duplicate tiles. Our own implementation doesn't quite do that yet :)

**Comment by jerstlouis** [comment]
@vincentsarago 

> is this defined in the spec?

I not 100% clear whether it is well stated in the spec, but my understanding is that the left-most tile is the one that exists i.e., integer division by the calescence factor, then multiplying again by the coalescence factor.
e.g., for GNOSIS Global Grid, the northernmost tiles are always power of two multiplied by 0, 1, 2 or 3.

>  3,0,1, is it the alias of 2,0,1 or the contrary?

col: 3, row: 0, level: 1, is an alias for col: 2, row: 0, level: 1 for the _GGG_.

**Comment by jerstlouis** [comment]
I don't think there is a formal requirement, but the following is included in normative [_Section 6.1.5_](https://docs.ogc.org/is/17-083r4/17-083r4.html#toc15):

> Even if tiles can coalesce, this does not change the indexing or the tile matrix set that will be the same as if no coalescence has been applied. For example, if the c
 coefficient is 4, the tileCol of the first tile will be 0, the tileCol of the second tile will be 4, the tileCol of the third tile will be 8 and so on. **In other words, and for the same example, tileCol 0, 1, 2 and 3 points to the same tile.**

**Comment by jerstlouis** [comment]
@vincentsarago The spec says they are the _same_ tile. It might be good to clarify that the first one should be the one that client look for, and that servers SHOULD redirect to the first one, and MAY return a 4xx for the non-canonical tile.
